src/policy/largeobjectspace.rs

```diff
@@ -86,6 +86,45 @@ impl<VM: VMBinding> SFT for LargeObjectSpace<VM> {
     fn is_mmtk_object(&self, addr: Address) -> bool {
         crate::util::metadata::vo_bit::is_vo_bit_set_for_addr::<VM>(addr).is_some()
     }
+    #[cfg(feature = "is_mmtk_object")]
+    fn find_object_from_internal_pointer(
+        &self,
+        ptr: Address,
+        max_search_bytes: usize,
+    ) -> Option<ObjectReference> {
+        use crate::util::metadata::vo_bit;
+        // For large object space, it is a bit special. We only need to check VO bit for each page.
+        let mut cur_page = ptr.align_down(BYTES_IN_PAGE);
+        let low_page = ptr
+            .saturating_sub(max_search_bytes)
+            .align_down(BYTES_IN_PAGE);
+        while cur_page > low_page {
+            // If the page start is not mapped, there can't be an object in it.
+            if !cur_page.is_mapped() {
+                return None;
+            }
+            if vo_bit::get_raw_vo_bit_word(cur_page) != 0 {
+                // Find the exact address that has vo bit set
+                for offset in 0..vo_bit::VO_BIT_WORD_TO_REGION {
```

Comment:
`VO_BIT_WORD_TO_REGION` is currently 512.  But what if the "in-object address" is 700 or 1000 bytes into the start of an object?  It is possible if the VM uses the Sable VM layout.

Reply:
It is very unlikely we will ever see an extreme case like that. However, I can try put in some assertions to make sure if that ever happens, we fail explicitly.

Reply:
I added an assertion when we set VO bit for LOS. If the VO bit is not in the first 512 bytes, we will panic.